DataServer/server.py

```diff
@@ -286,29 +297,74 @@ def __getRevision(self) -> str:
         self._log = logging.getLogger(self.__class__.__name__)
         self._log.info(f"Starting ASM Data Server v{DataServer.__version__}, {self.__getRevision()}")
         self.config = ServerConfig(config_file)
         devices_file = os.path.join(self.config.data_dir, 'devices.yaml')
         self.device_tree = devices.DeviceTree(devices_file)
         self.hostname = ''
         self.__client_queues: List[ClientHandler] = []
+        self._report_outage = report_outage
+
+        for device in self.device_tree.getDevices():
+            self.setup_device(device)
 
     async def run(self):
         self._log.info(f'Connecting to {self.hostname}:{self.config.port}')
+
         server = await asyncio.start_server(self.client_thread, self.hostname,
                                             self.config.port)
+        outage_detector = asyncio.create_task(self.start_outage_detector())
         async with server:
             await server.serve_forever()
 
+        outage_detector.cancel()
+
     async def client_thread(self, reader: StreamReader, writer: StreamWriter):
         client = ClientHandler(device_tree=self.device_tree, reader=reader,
-                               writer=writer, config=self.config)
+                               writer=writer, config=self.config,
+                               setup_device=self.setup_device)
         self.__client_queues.append(client)
         await client.run()
         self.__client_queues.remove(client)
 
+    def on_device_timeout(self, device: devices.Device):
+        if self._report_outage:
+            self._report_outage(device)
+        else:
+            self._log.warn(f'Device {device.deviceID} went offline but had no report outage hook')
+        device.offline = True
+
+    def on_device_connect(self, device: devices.Device):
+        device.offline = False
+
+    def setup_device(self, device: devices.Device):
+        device.on_timeout = self.on_device_timeout
+        device.on_connect = self.on_device_connect
+
+    async def start_outage_detector(self):
+        while True:
+            for device in self.device_tree.getDevices():
+                if device.last_comms is None:
+                    # The device may not have connected yet, or the device 
+                    # might never connect. Since we can't tell these cases apart
+                    # here, we will ignore it for now.
+                    continue
+
+                # No need to check outages on disconnected devices
+                if device.offline:
+                    continue
+
+                since_last_comms = dt.datetime.now() - device.last_comms
+                if since_last_comms.total_seconds() > self.config.heartbeat_timeout_secs:
+                    if device.on_timeout:
+                        device.on_timeout(device)
+                    else:
+                        self._log.warn(f'Device {device.deviceID} went offline but had no timeout hook')
```

Comment:
Again, let's try to use a generalized eventing mechanism.  We probably can also log here regardless of whether a timeout hook exists, and maybe also indicate how long it has been since we've heard from the node.